types/jest/index.d.ts

```diff
@@ -282,16 +282,21 @@ declare namespace jest {
      *   spy.mockRestore();
      * });
      */
-    function spyOn<T extends {}, M extends NonFunctionPropertyNames<Required<T>>>(
+    function spyOn<
+        T extends {},
+        Key extends keyof T,
+        A extends PropertyAccessors<Key, T> = PropertyAccessors<Key, T>,
+        Value extends Required<T>[Key] = Required<T>[Key],
+    >(
         object: T,
-        method: M,
-        accessType: 'get'
-    ): SpyInstance<Required<T>[M], []>;
-    function spyOn<T extends {}, M extends NonFunctionPropertyNames<Required<T>>>(
-        object: T,
-        method: M,
-        accessType: 'set'
-    ): SpyInstance<void, [Required<T>[M]]>;
+        method: Key,
+        accessType: A,
+    ):
+      A extends SetAccessor ? SpyInstance<void, [Value]>
+    : A extends GetAccessor ? SpyInstance<Value, []>
+    : Value extends Constructor ? SpyInstance<InstanceType<Value>, ConstructorArgsType<Value>>
+    : Value extends Func ? SpyInstance<ReturnType<Value>, ArgsType<Value>>
+    : never;
     function spyOn<T extends {}, M extends FunctionPropertyNames<Required<T>>>(
```

Comment:
@GerkinDev can you take another look? I refactored the return time to make it more readable, creating some new auxiliary types and letting the sequence of conditions a bit clearer.

Reply:
Way better, thank you :)